File: py/sorts/quicksort.py
# A quicksort
# https://en.wikipedia.org/wiki/Quicksort

# According to MS, this is wrong
# According to the Python Community, this is right
def recursiveQuicksort(values):
    if values:
        pivot, smaller, larger = values[0], [], []
        for item in values[1:]:
            if item < pivot:
                smaller.append(item)
            else:
                larger.append(item)
        return recursiveQuicksort(smaller) + [pivot] + recursiveQuicksort(larger)
    return values


# According to MS, this is right
# According to the Python Community, this is okay
# However it reeks of C
def inplaceQuicksort(values, begin=0, end=None):
    def partition(array, begin, end):
        pivot = begin
        for i in range(begin+1, end+1):
            if array[i] <= array[begin]:
                pivot += 1
                array[i], array[pivot] = array[pivot], array[i]
        array[pivot], array[begin] = array[begin], array[pivot]
        return pivot
    end = len(values) - 1 if end is None else end
    if begin >= end: return
    pivot = partition(array, begin, end)
    quicksort(array, begin, pivot-1)
    quicksort(array, pivot+1, end)

File: cs/sorts/quicksort.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Quicksort
{
    class Quicksort
    {

        static List<int> recursiveQuicksort(List<int> values)
        {
            if (values.Count() > 0)
            {
                var pivot = values[0];
                var smaller = new List<int>();
                var larger = new List<int>();
                for (int i = 1; i < values.Count(); i++)
                {
                    var item = values[i];
                    if (item < pivot)
                    {
                        smaller.Add(item);
                    } else
                    {
                        larger.Add(item);
                    }
                }

                var result = recursiveQuicksort(smaller);
                result.Add(pivot);
                result.AddRange(recursiveQuicksort(larger));

                return result;

            }
            return values;
        }
    }

}
